From: Jan Kara <jack@suse.cz>
Subject: scsi: Silence unnecessary warnings about ioctl to partition
References: bnc#758104
Patch-mainline: Never (upstream is happy with the warnings)

Sometimes, warnings about ioctls to partition happen often enough that they
form majority of the warnings in the kernel log and users complain. Since
the warnings are often for SG_IO ioctl itself we cannot get rid of them
completely (at least until userspace is verified) but we definitely can
stop warning when process has CAP_SYS_RAWIO capability and thus ioctl will
be successful.

Signed-off-by: Jan Kara <jack@suse.cz>

diff -rupX /crypted/home/jack/.kerndiffexclude linux-3.0-SLE11-SP2/block/scsi_ioctl.c linux-3.0-SLE11-SP2-1-ioctl_warn/block/scsi_ioctl.c
--- linux-3.0-SLE11-SP2/block/scsi_ioctl.c	2012-04-23 18:16:10.469185590 +0200
+++ linux-3.0-SLE11-SP2-1-ioctl_warn/block/scsi_ioctl.c	2012-04-23 21:27:39.712189308 +0200
@@ -722,11 +722,13 @@ int scsi_verify_blk_ioctl(struct block_d
 		break;
 	}
 
+	if (capable(CAP_SYS_RAWIO))
+		return 0;
+
 	/* In particular, rule out all resets and host-specific ioctls.  */
 	printk_ratelimited(KERN_WARNING
 			   "%s: sending ioctl %x to a partition!\n", current->comm, cmd);
-
-	return capable(CAP_SYS_RAWIO) ? 0 : -ENOTTY;
+	return -ENOTTY;
 }
 EXPORT_SYMBOL(scsi_verify_blk_ioctl);
 
